docs/books/valhalla-mjolnir/src/03_graph_construction_process.md

```diff
@@ -0,0 +1,310 @@
+# Graph Construction Process
+
+## Overview of the Graph Building Pipeline
+
+Mjolnir transforms OSM data into Valhalla's routing graph through a multi-stage pipeline. Understanding this pipeline is essential for building a graph tile builder. The main stages are:
+
+1. **Data Parsing**: Reading OSM data (covered in the previous chapter)
+2. **Initial Graph Construction**: Creating a basic graph structure
+3. **Graph Enhancement**: Adding additional attributes and metadata
+4. **Hierarchical Graph Building**: Creating multiple levels of the graph
+5. **Tile Creation**: Dividing the graph into tiles
+6. **Serialization**: Writing tiles to disk
+
+Let's examine each stage in detail.
+
+## Initial Graph Construction
+
+After parsing OSM data, Mjolnir constructs an initial graph using the `GraphBuilder` class:
+
+```cpp
+// From valhalla/mjolnir/graphbuilder.h
+class GraphBuilder {
+public:
+  static void Build(const boost::property_tree::ptree& pt,
+                   const std::string& ways_file,
+                   const std::string& way_nodes_file,
+                   const std::string& nodes_file,
+                   const std::string& edges_file,
+                   const std::string& complex_restriction_file,
+                   const std::string& osmdata_file);
+};
+```
+
+The implementation in `src/mjolnir/graphbuilder.cc` performs these key steps:
+
+1. **Node Creation**: Convert OSM nodes to graph nodes
+2. **Edge Creation**: Convert OSM ways to directed edges
+3. **Initial Attribution**: Assign basic attributes to edges (road class, surface type, etc.)
+4. **Connectivity Establishment**: Connect edges at nodes to form a network
+
+Here's a simplified version of the process:
```

Comment:
graph builder does indeed figure out the topological part of the data, ie which osm nodes should be become graph nodes and what edges connect to them but the main thing it does is it writes the initial tiling of the data at level 2, ie without hierarchies. the pseudo code below has a number of issues but chief among them is there is no concept of tiles and that we use a directed graph but always store an edge and its reversed direction edge regardless of accessibility wrt. direction (e.g. one way or otherwise we have 2 edges for each direction of travel between two graph nodes)

Reply:
> the pseudo code below has a number of issues but chief among them is there is no concept of tiles and that we use a directed graph but always store an edge and its reversed direction edge regardless of accessibility wrt. direction (e.g. one way or otherwise we have 2 edges for each direction of travel between two graph nodes)

Excellent note, I will update this section.